lib/api/metadataSearch.js

```diff
@@ -0,0 +1,159 @@
+    }
+    let validatedAst;
+    if (params.search !== undefined) {
+        const astOrError = validateSearchParams(params.search);
+        if (astOrError.error) {
+            return callback(astOrError.error);
+        }
+        validatedAst = astOrError.ast;
+    }
+    // AWS only returns 1000 keys even if max keys are greater.
+    // Max keys stated in response xml can be greater than actual
+    // keys returned.
+    const actualMaxKeys = Math.min(constants.listingHardLimit, requestMaxKeys);
+
+    const metadataValParams = {
+        authInfo,
+        bucketName,
+        requestType: 'metadataSearch',
+        request,
+    };
+    const listParams = {
+        listingType: 'DelimiterMaster',
+        maxKeys: actualMaxKeys,
+        delimiter: params.delimiter,
+        prefix: params.prefix,
+    };
+    if (v2) {
+        listParams.v2 = true;
+        listParams.startAfter = params['start-after'];
+        listParams.continuationToken =
+            decryptToken(params['continuation-token']);
+        listParams.fetchOwner = params['fetch-owner'] === 'true';
+    } else {
+        listParams.marker = params.marker;
+    }
+
+    metadataValidateBucket(metadataValParams, log, (err, bucket) => {
+        const corsHeaders = collectCorsHeaders(request.headers.origin,
+            request.method, bucket);
+        if (err) {
+            log.debug('error processing request', { error: err });
+            monitoring.promMetrics(
+                'GET', bucketName, err.code, 'metadataSearch');
+            return callback(err, null, corsHeaders);
+        }
+        if (params.versions !== undefined) {
+            listParams.listingType = 'DelimiterVersions';
+            delete listParams.marker;
+            listParams.keyMarker = params['key-marker'];
+            listParams.versionIdMarker = params['version-id-marker'] ?
+                versionIdUtils.decode(params['version-id-marker']) : undefined;
+        }
+        if (!requestMaxKeys) {
+            const emptyList = {
+                CommonPrefixes: [],
+                Contents: [],
+                Versions: [],
+                IsTruncated: false,
+            };
+            return handleResult(listParams, requestMaxKeys, encoding, authInfo,
+                bucketName, emptyList, corsHeaders, log, callback);
+        }
+        if (params.search !== undefined) {
+            log.info('performing search listing', { search: params.search });
+            try {
+                listParams.mongifiedSearch = parseWhere(validatedAst);
+            } catch (err) {
+                log.debug(err.message, {
+                    stack: err.stack,
+                });
+                monitoring.promMetrics(
+                    'GET', bucketName, 400, 'metadataSearch');
+                return callback(errors.InvalidArgument
+                    .customizeDescription('Invalid sql where clause ' +
+                        'sent as search query'));
+            }
+        }
```

Comment:
Because here we are always running this code, since the actual function will only be called with `params.search` set, then maybe we can move this block of code above, so that we don't do unnecessary calculations if the `parseWhere` throws an error? (Maybe not, just suggesting!)

Reply:
solved https://github.com/scality/cloudserver/pull/4224/commits/826e7bd7b4ad400430f80a5eec1d6651b1df38dc